src/odb/src/db/dbBlock.cpp

```diff
@@ -2089,7 +2098,65 @@ void dbBlock::setDieArea(const Rect& new_area)
   }
 }
 
+void dbBlock::setDieArea(const Polygon& new_area)
+{
+  _dbBlock* block = (_dbBlock*) this;
+  block->_die_area = new_area;
+  for (auto callback : block->_callbacks) {
+    callback->inDbBlockSetDieArea(this);
+  }
+
+  dbTech* tech = getTech();
+  dbSet<dbTechLayer> tech_layers = tech->getLayers();
+  // General flow here:
+  // We have a polygon floorplan. We can represent this floorplan
+  // as a bounding rectangle with obstructions where there is empty
+  // space in the polygon.
+  //
+  // The following code runs a subtraction on the bounding box and
+  // the polygon. This generates a list of polygons where there is
+  // supposed to be empty space. These polygons are then decomposed
+  // into rectangles. Which are then used to create obstructions and
+  // placement blockages.
+  std::list<Polygon> results;
+  Polygon bounding_rect = block->_die_area.getEnclosingRect();
+  boost::geometry::difference(bounding_rect, block->_die_area, results);
+  for (odb::Polygon& blockage_area : results) {
+    std::vector<Rect> blockages;
+    decompose_polygon(blockage_area.getPoints(), blockages);
+    for (odb::Rect& blockage_rect : blockages) {
+      dbBlockage* db_blockage = dbBlockage::create(this,
+                                                   blockage_rect.xMin(),
+                                                   blockage_rect.yMin(),
+                                                   blockage_rect.xMax(),
+                                                   blockage_rect.yMax());
+      db_blockage->setIsVirtual(true);
+
+      // Create routing blockages
+      for (dbTechLayer* tech_layer : tech_layers) {
+        if (tech_layer->getType() == odb::dbTechLayerType::OVERLAP) {
+          continue;
+        }
+
+        dbObstruction* obs = dbObstruction::create(this,
+                                                   tech_layer,
+                                                   blockage_rect.xMin(),
+                                                   blockage_rect.yMin(),
+                                                   blockage_rect.xMax(),
+                                                   blockage_rect.yMax());
+        obs->setIsVirtual(true);
+      }
+    }
+  }
+}
+
 Rect dbBlock::getDieArea()
+{
+  _dbBlock* block = (_dbBlock*) this;
+  return block->_die_area.getEnclosingRect();
```

Comment:
I think this should become `getDieBBox()` and `getDieAreaPolygon()` should be `getDieArea()`.

Reply:
After talking with Peter below I think you're right. I'll change it.

Reply:
Do you mind if I send this in a follow up PR. It modified about 100 files, and I think would be better as a stand alone PR.